cmd/yggd/main.go

```diff
@@ -370,51 +371,79 @@ func createTransport(c *cli.Context, tlsConfig *tls.Config, d *dispatcher) (tran
 	}
 }
 
-func handleControlMessage(msg []byte, t transport.Transport) {
-	var cmd yggdrasil.Command
-	if err := json.Unmarshal(msg, &cmd); err != nil {
-		log.Errorf("cannot unmarshal control message: %v", err)
-		return
-	}
-
-	log.Debugf("received message %v", cmd.MessageID)
-	log.Tracef("command: %+v", cmd)
-	log.Tracef("Control message: %v", cmd)
-
-	switch cmd.Content.Command {
-	case yggdrasil.CommandNamePing:
-		event := yggdrasil.Event{
-			Type:       yggdrasil.MessageTypeEvent,
-			MessageID:  uuid.New().String(),
-			ResponseTo: cmd.MessageID,
-			Version:    1,
-			Sent:       time.Now(),
-			Content:    string(yggdrasil.EventNamePong),
+func createControlMessageHandler(d *dispatcher) func(msg []byte, t transport.Transport) {
+	return func(msg []byte, t transport.Transport){
+		var cmd yggdrasil.Command
+		if err := json.Unmarshal(msg, &cmd); err != nil {
+			log.Errorf("cannot unmarshal control message: %v", err)
+			return
 		}
 
-		err := t.SendControl(event)
-		if err != nil {
-			log.Error(err)
-		}
-	case yggdrasil.CommandNameDisconnect:
-		log.Info("disconnecting...")
-		t.Disconnect(500)
-	case yggdrasil.CommandNameReconnect:
-		log.Info("reconnecting...")
-		t.Disconnect(500)
-		delay, err := strconv.ParseInt(cmd.Content.Arguments["delay"], 10, 64)
-		if err != nil {
-			log.Errorf("cannot parse data to int: %v", err)
-			return
+		log.Debugf("received message %v", cmd.MessageID)
+		log.Tracef("command: %+v", cmd)
+		log.Tracef("Control message: %v", cmd)
+
+		switch cmd.Content.Command {
+		case yggdrasil.CommandNamePing:
+			event := yggdrasil.Event{
+				Type:       yggdrasil.MessageTypeEvent,
+				MessageID:  uuid.New().String(),
+				ResponseTo: cmd.MessageID,
+				Version:    1,
+				Sent:       time.Now(),
+				Content:    string(yggdrasil.EventNamePong),
+			}
+
+			err := t.SendControl(event)
+			if err != nil {
+				log.Error(err)
+			}
+		case yggdrasil.CommandNameDisconnect:
+			log.Info("disconnecting...")
+			disconnectAllWorkers(d)
+			t.Disconnect(500)
+
+		case yggdrasil.CommandNameReconnect:
+			log.Info("reconnecting...")
+			t.Disconnect(500)
+			delay, err := strconv.ParseInt(cmd.Content.Arguments["delay"], 10, 64)
+			if err != nil {
+				log.Errorf("cannot parse data to int: %v", err)
+				return
+			}
+			time.Sleep(time.Duration(delay) * time.Second)
+
+			if err := t.Start(); err != nil {
+				log.Errorf("cannot reconnect to broker: %v", err)
+				return
+			}
+		default:
+			log.Warnf("unknown command: %v", cmd.Content.Command)
 		}
-		time.Sleep(time.Duration(delay) * time.Second)
+	}
 
-		if err := t.Start(); err != nil {
-			log.Errorf("cannot reconnect to broker: %v", err)
-			return
+}
+
+func disconnectAllWorkers(d *dispatcher){
+	for _, w := range d.workers {
+		f := func() {
```

Comment:
That's not exactly what I had in mind. Please just extract the body of the for loop to a separate function and call it in the loop.